### ISSUE TITLE
[Assembler] Yul code compiled for EVM <= byzantium returns invalid instruction status code on EVMC call

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#12590]
```
{
  function f() {
    sstore(sload(caller()), exp(0xffffffffff,0xffff))
  }
  f()
  f()
  f()
}
```

fails https://github.com/ethereum/solidity/blob/597426bd1c93e1d75940d5e9dacdb55123228a01/test/tools/ossfuzz/StackReuseCodegenFuzzer.cpp#L99-L102

To reproduce, one would have to assemble the Yul code targeted at tangerine whistle, deploy and issue a call and print status code of the call.

**Comment by bshastry** [comment]
The question here is why does the assembler emit an invalid opcode for tangerine whistle. Is the `exp` opcode not supported for that EVM version?

**Comment by bshastry** [comment]
Invalid instruction is generated for EVM versions `<= byzantium`

**Comment by bshastry** [comment]
Invalid instruction is also generated for latest EVM version (london). For example:

```
{
  function g() -> o {
    sstore(0, exp(0x1fff,0xfff))
  }
  function f() {
    pop(g())
  }
  f()
  f()
}
```

**Comment by chriseth** [comment]
The only thing I could think of is that the optimizer adds a shift opcode, but I currently cannot reproduce this.

Maybe @gumb0 has any insights into this?

**Comment by gumb0** [comment]
Yes, shifts were added in Constantinople, so shift opcode in Byzantium would be invalid.

As for London no idea, nothing was deprecated for sure.
But that (london) code should generate stack underflow instead? (could be wrong error from evmone?)

**Comment by chriseth** [comment]
Yeah, I don't think the shift opcode is the problem, I think it's rather either a problem in evmc or how we interface with evmc.
